### ISSUE TITLE
Namespace and element name are switched

### ISSUE THREAD

**Issue by licaon-kter** [processone/ejabberd#2744]
ejabberd 18.09 iirc (@vanitasvitae right?)

Dino HEAD, initial error log: https://gist.github.com/licaon-kter/c6957c9886a10e62f57a85884c125d85

Using the branch `fiaxh/debug_reconnect_ns_error`:

`** (dino:26464): WARNING **: 16:10:28.164: stanza_reader.vala:262: XmlError at: <message to='user@domain.tld/dino.ea2fed25' from='user@domain.tld'><result id='1544819733897019' xmlns='urn:xmpp:mam:2'><forwarded xmlns='urn:xmpp:forward:0'><message to='user@domain.tld/dino.ea2fed25' from='user2@domain2.tld' id='8213063422210396831' xmlns='jabber:client'><archived by='user@domain.tld' id='1544819733897019' xmlns='urn:xmpp:mam:tmp'/><stanza-id by='user@domain.tld' id='1544819733897019' xmlns='urn:xmpp:sid:0'/><jabber:x:oob xmlns='x'><url>https://domain2.tld:5280/captcha/8213063422210396831</url></jabber:x:oob><urn:xmpp:captcha xmlns='captcha'><jabber:x:data xmlns='x' type='form'><field var='FORM_TYPE' type='hidden'><value>urn:xmpp:captcha</value></field><field var='from' type='hidden'><value>user@domain.tld</value></field><field var='challenge' type='hidden'><value>8213063422210396831</value></field><field var='sid' type='hidden'><value/></field><field var='ocr' type='text-single' label='Enter the text you see'><urn:xmpp:media-element xmlns='media'><uri type='image/png'>cid:sha1+snip@bob.xmpp.org</uri></urn:xmpp:media-element><required/></field></jabber:x:data></urn:xmpp:captcha><urn:xmpp:bob xmlns='data' type='image/png' max-age='0' cid='sha1+snip@bob.xmpp.org'>something hashy, long that ends up suddenly without closing brackets`

An then it loops the log as previously linked.

So, it's an old captcha request that triggers this?
 
> Ah, now I got the issue. Not the namespace itself is disallowed, but namespace and element name are switched. If this is still happening in latest ejabberd, it is probably worth being reported to them.

Ref: https://github.com/dino/dino/issues/394#issuecomment-450594710

**Comment by horazont** [comment]
Maybe that one https://github.com/siacs/Conversations/issues/3315 is another instance of this?

**Comment by vanitasvitae** [comment]
Yeah, looks like it.

**Comment by prefiks** [comment]
Are you guys using mam with mysql backend? If so you probably need to modify db schema of archive/spool table to make it look like in latest version. Previously it allowed storing max 64k of data, and mysql was silently cutting data past that. After schema update it would allow storing 16M.

**Comment by Flowdalic** [comment]
@Zinid: Yes the report does not actually mention and makes it hard to spot the actual issue.

It appears that ejabberd sends:

```
<urn:xmpp:media-element xmlns='media'>
```

**Comment by badlop** [comment]
How to reproduce the problem?

I installed ejabberd 18.09, enabled CAPTCHA and MAM, keep mnesia storage, created a room that requires CAPTCHA, try to join the room with another account, so it receives the CAPTCHA form message that gets stored in the MAM mnesia table. Then the user queries the archived messages, and the CAPTCHA one is returned correctly, including:
```xmpp
<media xmlns='urn:xmpp:media-element'>
  <uri type='image/png'>cid:sha1+c7314c89aed7...@bob.xmpp.org</uri>
</media>
```

**Comment by vanitasvitae** [comment]
> How to reproduce the problem?

I dont know the answer unfortunately.
I'm using MAM using a mysql/mariadb database, so maybe @prefiks conment is a hint?

Maybe the stanzas in question got messed up due to a bug in ejabberds MAM implementation in conjunction with mysql?

**Comment by prefiks** [comment]
I didn't spot that wrong element, so i though the problem was with data that was fetched from mam storage was missing closing tags (like when content was cut), but this will not lead to having ns/name switched, so there must be different reason for that

**Comment by licaon-kter** [comment]
How do I test exactly? What (and how) do I extract the needed info from my (server) side? (Since Dino crashes for me with that data from vanitasvitae)

FYI, I'm on postgres

**Comment by badlop** [comment]
Some ideas to try in the server where the problem appears:
* Check that messed stanza in the mam table: is it stored correctly (and messed later when sent to the user), or is it stored already messed?
* Try to join such a room with another account, then query MAM: does the problem still reproduce (the problem appears with current software), or it works ok now (the problem was provoked some time ago, maybe with old sorftware, and there's just a few messed messages in the database)

**Comment by vanitasvitae** [comment]
> Check that messed stanza in the mam table: is it stored correctly (and messed later when sent to the user), or is it stored already messed?

How to check that?

**Comment by licaon-kter** [comment]
@zinid 

* domain1 mine, HEAD, compressed_xml, user1 me, [this config](https://gist.github.com/licaon-kter/094dca8262218474ef34e2a349995ba6)
* domain2 @vanitasvitae, 18.09, normal stanzas, user2 him

At one point I contacted him for the first time, got the captcha message, the first one did not do anything, or maybe he wasn't paying attention, doesn't matter, got a second one, we talked for a bit back in December.

Now, at the end of December some new commits on Dino were made, I updated & recompiled, started and I got (still get) stuck in a loop, every time it goes like this: https://github.com/dino/dino/issues/394#issuecomment-450520477

@badlop: I've dumped the DB, grepped for captcha and got this: https://gist.github.com/licaon-kter/1c8c620be5cb065b0b80503dbaf87272 (use line wrap)

But is that what you wanted?

**Comment by prefiks** [comment]
@licaon-kter 

Could you be able to start ejabberd debug console (execute "ejabberdctl debug" in shell), and when you get to erlang prompt execute this line:

io:format("~p", [ejabberd_sql:sql_query(<<"domain1.tld">>, <<"select xml from archive where timestamp=1544819733897019">>)]).

and post somewhere result that this generate? Change domain1.tld if you are using something else.

**Comment by prefiks** [comment]
This should be fixed by commit 06d69c5277f08d9a57f74836c247d3927ca390e6

**Comment by aveao** [comment]
Can confirm that the fix works.
